File: app/src/main/java/com/github/goldy1992/mp3player/service/player/AudioBecomingNoisyBroadcastReceiver.java
package com.github.goldy1992.mp3player.service.player;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.media.AudioManager;

import androidx.annotation.VisibleForTesting;

import com.google.android.exoplayer2.ExoPlayer;

import javax.inject.Inject;

public class AudioBecomingNoisyBroadcastReceiver extends BroadcastReceiver {

    private final ExoPlayer player;
    private final Context context;
    private boolean audioNoisyReceiverRegistered = false;

    /**O
     * Constructor
     */
    @Inject
    public AudioBecomingNoisyBroadcastReceiver(Context context, ExoPlayer exoPlayer) {
        this.context = context;
        this.player = exoPlayer;
    }
    @Override
    public synchronized void onReceive(Context context, Intent intent) {
        final boolean intentIsAudioBecomingNoisy =
                AudioManager.ACTION_AUDIO_BECOMING_NOISY.equals(intent.getAction());
        if (intentIsAudioBecomingNoisy) {
            player.setPlayWhenReady(false); // pause the player
        }
    }

    public void register() {
        if (!isRegistered()) {
            IntentFilter audioNoisyIntentFilter =
                    new IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY);
            context.registerReceiver(this, audioNoisyIntentFilter);
            audioNoisyReceiverRegistered = true;
        }
    }

    public void unregister() {
        if (isRegistered()) {
            context.unregisterReceiver(this);
            audioNoisyReceiverRegistered = false;
        }
    }

    @VisibleForTesting
    public boolean isRegistered() {
        return audioNoisyReceiverRegistered;
    }
}

File: service/src/main/java/com/github/goldy1992/mp3player/service/player/AudioBecomingNoisyBroadcastReceiver.kt
package com.github.goldy1992.mp3player.service.player

import android.Manifest
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioManager
import android.os.Handler
import androidx.annotation.VisibleForTesting
import androidx.media3.exoplayer.ExoPlayer

import dagger.hilt.android.qualifiers.ApplicationContext
import javax.inject.Inject

class AudioBecomingNoisyBroadcastReceiver
/**
 * Constructor
 */ @Inject constructor(@ApplicationContext private val context: Context, private val player: ExoPlayer) : BroadcastReceiver() {
    @get:VisibleForTesting
    var isRegistered = false
        private set

    @Synchronized
    override fun onReceive(context: Context, intent: Intent) {
        val intentIsAudioBecomingNoisy = AudioManager.ACTION_AUDIO_BECOMING_NOISY == intent.action
        if (intentIsAudioBecomingNoisy) {
            player.playWhenReady = false // pause the player
        }
    }

    fun register() {
        if (!isRegistered) {
            val audioNoisyIntentFilter = IntentFilter(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
            context.registerReceiver(this, audioNoisyIntentFilter, Manifest.permission.MODIFY_AUDIO_SETTINGS,null)
            isRegistered = true
        }
    }

    fun unregister() {
        if (isRegistered) {
            context.unregisterReceiver(this)
            isRegistered = false
        }
    }

}